nbcelltests/tests_vendored.py

```diff
@@ -58,35 +58,94 @@
 
 
 class TestNotebookBase(unittest.TestCase):
-    # abstract - subclasses must define KERNEL_NAME
+    """Base class for representing a notebook's code cells and their
+    associated tests; can submit cells to a kernel, track which cells
+    have been executed, and ensure all necessary cells are executed in
+    order.
+
+    For instance, requesting to run test_code_cell_7,
+    test_code_cell_8, and test_code_cell_9 (in that order) will result
+    in:
+
+      1. test_code_cell_7: executes cells 1, 2, 3, 4, 5, 6, 7
+      2. test_code_cell_8: executes cell 8
+      3. test_code_cell_9: executes cell 9
+
+    The above assumes all cells+tests succeed.
+
+    If e.g. cell 3 fails, the above changes to:
+
+      1. test_code_cell_7: executes cells 1, 2; fails on 3 (with a
+                           message that cell 3 failed)
+      2. test_code_cell_8: also fails on 3
+      3. test_code_cell_9: also fails on 3
+
+    Requesting to run test_code_cell_5 and test_code_cell_3 (in that order)
+    will result in:
+
+      1. test_code_cell_5: executes cells 1, 2, 3, 4, 5 (test passes)
+      2. test_code_cell_3: execute nothing (test passes)
+
+    This is an abstract class; subclasses will supply the source of
+    code cells and their associated tests in cells_and_tests, plus
+    test methods as entry points for test runners.
+
+    Note: 'cell' used in this class refers to cell number; 'cell
+    content' typically refers to code_cell+test (depending what is
+    passed in).
+    """
+    # abstract - subclasses will define KERNEL_NAME (TODO: make
+    # actually abstract...)
 
     @classmethod
     def setUpClass(cls):
+        # TODO: these probably ought to be in __init__ - want one per
```

Comment:
I think `setUpClass` might actually be better here, as it delays the startup as long as possible (e.g. if all tests are marked as skipped, this won't be run).

Reply:
My thinking was: Is there ever a case during typical test running that there can be multiple instances of a test class (in the same thread)? That's what I am unsure of. If not, I agree with you. If so, I would prefer the safety of doing it `__init__`.

Reply:
I think the answer here is pretty significant: https://stackoverflow.com/questions/18056464/when-should-i-use-setupclass-and-when-init .

> They are also called at different times. The test runner creates a new instance for every test. If you test class contains 5 test methods, 5 instances are created and __init__ is called 5 times.

Also: It says that if the tests are shuffled, `tearDownClass` will always be called before `setUpClass` is called the second time.

Reply:
Ok, that answers that then. Thanks!